storage/innobase/lock/lock0lock.cc

```diff
@@ -4692,61 +4768,90 @@ void lock_release_on_prepare(trx_t *trx)
+       lock= prev)
   {
-    LockMutexGuard g{SRW_LOCK_CALL};
-    trx->mutex_lock();
-
-    for (lock_t *prev, *lock= UT_LIST_GET_LAST(trx->lock.trx_locks); lock;
-         lock= prev)
+    ut_ad(lock->trx == trx);
+    prev= UT_LIST_GET_PREV(trx_locks, lock);
+    if (!lock->is_table())
     {
-      ut_ad(lock->trx == trx);
-      prev= UT_LIST_GET_PREV(trx_locks, lock);
-      if (!lock->is_table())
+      ut_ad(!lock->index->table->is_temporary());
+      if (!lock->is_rec_granted_exclusive_not_gap())
+        lock_rec_dequeue_from_page(lock, false);
+      else if (UNIV_UNLIKELY(lock->type_mode &
+                             (LOCK_PREDICATE | LOCK_PRDT_PAGE)))
+        /* SPATIAL INDEX locking is broken. */;
+      else
       {
-        ut_ad(!lock->index->table->is_temporary());
-        if (!lock->is_rec_granted_exclusive_not_gap())
-          lock_rec_dequeue_from_page(lock, false);
-        else if (UNIV_UNLIKELY(lock->type_mode &
-                               (LOCK_PREDICATE | LOCK_PRDT_PAGE)))
-          /* SPATIAL INDEX locking is broken. */;
-        else
+        auto cell= lock_sys.rec_hash.cell_get(lock->un_member.rec_lock.
+                                              page_id.fold());
+        if (lock_rec_get_nth_bit(lock, PAGE_HEAP_NO_SUPREMUM))
+          lock_rec_unlock(*cell, lock, PAGE_HEAP_NO_SUPREMUM);
+        else if (unlock_unmodified)
         {
-          auto cell= lock_sys.rec_hash.cell_get(lock->un_member.rec_lock.
-                                                page_id.fold());
-          if (lock_rec_get_nth_bit(lock, PAGE_HEAP_NO_SUPREMUM))
-            lock_rec_unlock(*cell, lock, PAGE_HEAP_NO_SUPREMUM);
-          else
-          {
-            ut_ad(lock->trx->isolation_level > TRX_ISO_READ_COMMITTED ||
-                  /* Insert-intention lock is valid for supremum for isolation
-                     level > TRX_ISO_READ_COMMITTED */
-                  lock->mode() == LOCK_X ||
-                  !lock_rec_get_nth_bit(lock, PAGE_HEAP_NO_SUPREMUM));
-            lock_rec_unlock_unmodified(*cell, lock, offsets, heap, mtr);
+          ut_ad(lock->trx->isolation_level > TRX_ISO_READ_COMMITTED ||
+                /* Insert-intention lock is valid for supremum for isolation
+                   level > TRX_ISO_READ_COMMITTED */
+                lock->mode() == LOCK_X ||
+                !lock_rec_get_nth_bit(lock, PAGE_HEAP_NO_SUPREMUM));
+          ulint set_nth_bit_calls= trx->lock.set_nth_bit_calls;
+          lock_sys.wr_unlock();
+          trx->mutex_unlock();
+          mtr.start();
+          buf_block_t *block= btr_block_get(
+              *lock->index, lock->un_member.rec_lock.page_id.page_no(),
+              RW_S_LATCH, true, &mtr, nullptr, nullptr
+#if defined(UNIV_DEBUG) || !defined(DBUG_OFF)
+              , BUF_GET_POSSIBLY_FREED
+#endif /* defined(UNIV_DEBUG) || !defined(DBUG_OFF) */
+          );
+          lock_sys.wr_lock(SRW_LOCK_CALL);
+          if (block) {
```

Comment:
In the common MariaDB formatting style, Monty wants us to deviate from K&R and split the line before `{` except when it is after a `switch` expression.